[ConstantRange][LVI] Add initial support for `multiplyWithNoWrap`

antoniofrighetto opened the pull request:
Introduce support for computing multiplication ranges when nowrap flags are known. This is achieved by intersecting the multiplication range with the saturating one. Note that we may still conservatively return overdefined when handling non-wrapped/non-sign-wrapped ranges.

antoniofrighetto committed:
[ConstantRange][LVI] Add support for `multiplyWithNoWrap`

Introduce support for computing multiplication ranges when nowrap
flags are known. This is achieved by intersecting the multiplication
range with the saturating one. Note that we may still conservatively
return overdefined when handling non-wrapped/non-sign-wrapped ranges.

dtcxzyw commented on llvm/lib/IR/ConstantRange.cpp:
```diff
@@ -1167,6 +1169,28 @@ ConstantRange::multiply(const ConstantRange &Other) const {
+                                  PreferredRangeType RangeType) const {
+  if (isEmptySet() || Other.isEmptySet())
+    return getEmpty();
+
```
It would be better to early exit when both ranges are full.

dtcxzyw commented on llvm/lib/IR/ConstantRange.cpp:
```diff
@@ -1167,6 +1169,28 @@ ConstantRange::multiply(const ConstantRange &Other) const {
+  // cases. To be extended.
+
+  if (NoWrapKind & OverflowingBinaryOperator::NoSignedWrap)
+    Result = Result.intersectWith(smul_sat(Other), RangeType);
```
@nikic Can you test the compile-time impact?

XChy commented on llvm/lib/IR/ConstantRange.cpp:
```diff
@@ -1167,6 +1169,28 @@ ConstantRange::multiply(const ConstantRange &Other) const {
+
+  ConstantRange Result = multiply(Other);
+
+  // TODO: Bounds are not yet precise if the input sets are non-wrapped /
```
Could you please provide a negative test or an example? It's hard for me to imagine it.

antoniofrighetto replied:
For example, given input ranges `[0,2)` and `[2,0)` (bitwidth: 4), we currently return `full-set`, but a more refined CR would be `[2,1)`.

antoniofrighetto commented:
TBH I have been having a hard time understanding some of the failures that arise when testing exhaustively mul nuw w/ optimality `CheckNonWrappedOnly`. For instance, for input ranges `[0,5)`, `[6,-1)` (bitwidth 4), I get the following:
```
Failure
Value of: NotPreferred(PossibleCR)
  Actual: false (Inputs = [0,5), [6,-1), CR = full-set, BetterCR = [6,1))
Expected: true

Failure
Value of: NotPreferred(PossibleCR)
  Actual: false (Inputs = [0,5), [6,-1), CR = full-set, BetterCR = [0,-1))
Expected: true
```
Isn't the union of the two sets full-set? What am I missing?

nikic commented:
@antoniofrighetto 0,1,2,3,4 *nuw 6,7,8,9,10,11,12,13,14 = 0,6,7,8,9,10,11,12,13,14

I don't think having accurate mul ranges is feasible without a brute force evaluation. You can drop the TODO.

nikic commented on llvm/test/Transforms/SCCP/range-mul-nuw-nsw-flags.ll:
```diff
@@ -0,0 +1,30 @@
+  %cond = icmp ult i32 %add, 4999
+  ret i1 %cond
+else:
+  call void @opaque()
```
This call is unnecessary.

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
+  EXPECT_EQ(ConstantRange(APInt(4, 0), APInt(4, 2))
+                .multiplyWithNoWrap(ConstantRange(APInt(4, 2), APInt(4, 0)),
+                                    OBO::NoUnsignedWrap),
+            ConstantRange(4, true));
```
```suggestion
            ConstantRange(4));
```
Not signed.

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
+  EXPECT_EQ(Full.multiplyWithNoWrap(Some, OBO::NoSignedWrap), Full);
+  EXPECT_EQ(Some.multiplyWithNoWrap(Full, OBO::NoSignedWrap), Full);
+  EXPECT_EQ(ConstantRange(APInt(4, 0), APInt(4, 4))
+                .multiplyWithNoWrap(ConstantRange(APInt(4, -5), APInt(4, 4)),
```
```suggestion
                .multiplyWithNoWrap(ConstantRange(APInt(4, -5, true), APInt(4, 4)),
```
But this is... (also below).

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
+
+  auto CheckCorrectnessOnly = [](const ConstantRange &, const ConstantRange &) {
+    return false;
+  };
```
I'd move this to the top where the others are defined.

nikic commented on llvm/unittests/IR/ConstantRangeTest.cpp:
```diff
@@ -1019,6 +1019,99 @@ TEST_F(ConstantRangeTest, Multiply) {
+      [](const APInt &N1, const APInt &N2) -> std::optional<APInt> {
+        bool IsOverflow;
+        APInt Res = N1.umul_ov(N2, IsOverflow);
+        if (IsOverflow) {
```
No braces

antoniofrighetto commented:
Argh, for some reason, I kept on considering them as if they were two different ranges :( Addressed reviews, thanks! (The `(4, true)` instance uses the CR constructor for empty/full set, not the APInt one)

dtcxzyw reviewed (approved):
LGTM.

nikic reviewed (approved):
LGTM
